src/dos/program_ls.cpp

```diff
@@ -22,14 +22,240 @@
-	safe_strcpy(args, tmp.c_str());
-	first_shell->CMD_LS(args);
+	// Handle command line
+
+	if (HelpRequested()) {
+		MoreOutputStrings output(*this);
+		output.AddString(MSG_Get("PROGRAM_LS_HELP_LONG"));
+		output.Display();
+		return;
+	}
+
+	constexpr bool remove_if_found = true;
+	const bool has_option_all = cmd->FindExist("/a", remove_if_found);
+
+	std::vector<std::string> patterns = {};
+	cmd->FillVector(patterns);
+
+	// Make sure no other switches are supplied
+
+	std::string tmp_str = {};
+	if (cmd->FindStringBegin("/", tmp_str)) {
+		tmp_str = std::string("/") + tmp_str;
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH"), tmp_str.c_str());
+		return;
+	}
+
+	// Prepare search parameters
+
+	FatAttributeFlags search_attr = FatAttributeFlags::NotVolume;
+	if (!has_option_all) {
+		search_attr.system = false;
+		search_attr.hidden = false;
+	}
+
+	if (patterns.empty()) {
+		patterns.push_back({});
+	}
+	for (auto& pattern : patterns) {
+		pattern = to_search_pattern(pattern.c_str());
+	}
+
+	// Search for files/directories according to pattern
+
+	std::vector<DOS_DTA::Result> dir_contents = {};
+
+	const RealPt original_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	DOS_DTA dta(dos.dta());
+
+	for (const auto& pattern : patterns) {
+		if (shutdown_requested) {
+			break;
+		}
+
+		if (!DOS_FindFirst(pattern.c_str(), search_attr)) {
+			continue;
+		}
+
+		do {
+			DOS_DTA::Result result = {};
+			dta.GetResult(result);
+			if (result.IsDummyDirectory()) {
+				continue;
+			}
+			dir_contents.push_back(result);
+		} while (!shutdown_requested && DOS_FindNext());
+	}
+
+	dos.dta(original_dta);
+	if (shutdown_requested) {
+		return;
+	}
+
+	if (dir_contents.empty()) {
+		WriteOut(MSG_Get("SHELL_NO_FILES_SUBDIRS_TO_DISPLAY"));
+		return;
+	}
+
+	// Sort directory contents
+
+	constexpr bool reverse_order = false;
+	DOS_Sort(dir_contents,
+	         ResultSorting::ByName,
+	         reverse_order,
+	         ResultGrouping::NonFilesFirst);
+
+	// Remove duplicates (might arise from multiple patterns)
+
+	const auto tmp = std::move(dir_contents);
+	dir_contents.clear();
+	for (const auto& entry : tmp) {
+		if (dir_contents.empty() ||
+		    dir_contents.back().IsDirectory() != entry.IsDirectory() ||
+		    dir_contents.back().name != entry.name) {
+			dir_contents.push_back(entry);
+		}
+	}
+
+	// Display search results
+
+	constexpr uint8_t separation = 2; // chars separating columns
+
+	const auto name_widths   = GetFileNameLengths(dir_contents, separation);
+	const auto column_widths = GetColumnWidths(name_widths, separation + 1);
+
+	const size_t num_columns = column_widths.size();
+
+	constexpr int ansi_blue  = 34;
+	constexpr int ansi_green = 32;
+
+	auto write_color = [&](const int ansi_color,
+	                       const std::string& txt,
+	                       const uint8_t width) {
+		const auto margin_size = static_cast<int>(width - txt.size());
+		WriteOut("\033[%d;1m%s\033[0m%-*s", ansi_color, txt.c_str(), margin_size, "");
```

Comment:
> ```c++
> constexpr int ansi_blue  = 34;
> constexpr int ansi_green = 32;
> WriteOut("\033[%d;1m%s\033[0m% ...);
> ```

We can make use of the `convert_ansi_markup()` routine:


```c++
auto write_color = [&](const std::string& colour,
                       const std::string& txt,
                       const uint8_t width) {
	const auto ansi_txt = convert_ansi_markup(colour + txt + "[reset]");
	const auto margin_size = static_cast<int>(width - txt.size());
	WriteOut("%s%-*s", ansi_txt.c_str(), margin_size, "");
};
```

To replace the  to replace the ANSI numbers and escape sequences with the colour tags:

```c++
write_color("[color=light-blue]", name, column_width);
write_color("[color=light-green]", name, column_width);
```

Reply:
> `const auto margin_size = static_cast<int>(width - txt.size());`

Is there any risk of this going negative?

Reply:
Changed the code to use `convert_ansi_markup` - but I'm calling them before the loop starts, to generate `ansi_blue`, `ansi_green` and `ansi_reset` as strings just once.

We shouldn't go negative (or even 0), unless there is a bug in `LS::GetFileNameLengths` or `LS::GetColumnWidths`. Added an assertion there.